### ISSUE TITLE
Reproducing DIRE Results on GenImage: Discrepancy Between ADM and SDv1 as Reconstruction Model

### ISSUE THREAD

**Issue by Ck2125** [beibuwandeluori/DRCT#13]
I saw in the paper that "Specifically, for the training of DIRE and DRCT, the reconstruction model is also SDv1." in the Comparison on GenImage section. However, the original DIRE paper and the public github code use ADM as the reconstruction model. Have you re-modified the code of compute_dire?
I encountered a problem when reproducing the DIRE results on the genimage dataset in your paper on the sdv1.4 subset of genimage. The results showed that the accuracy on the validation set of the eight subclasses of genimage was very high, close to 1. The reconstruction model I used is the compute_dire.py published by DIRE. I think there is a problem here (it is inconsistent with the statement in your paper). I would like to ask how you handled it?

**Comment by beibuwandeluori** [comment]
def calculate_sdie(img, sdir_img, is_success=True, input_size=224, phase='train'):
    if not is_success:
        return torch.zeros(size=(3, input_size, input_size), dtype=torch.float32)
    sdie_transforms = create_sdie_transforms(size=input_size, phase=phase)

    data = sdie_transforms(image=img, rec_image=sdir_img)
    img, sdir_img = data['image'], data['rec_image']

    # norm [0,255] -> [-1, 1]
    img = img / 127.5 - 1
    sdir_img = sdir_img / 127.5 - 1
    # absolute error
    sdie = np.abs(img - sdir_img)
    sdie_tensor = torch.from_numpy(np.transpose(np.array(sdie, dtype=np.float32), [2, 0, 1]))

    return sdie_tensor

This is the code we used to reproduce the DIRE calculations. We used the SDv1-inpainting and SDv2-inpainting models for image reconstruction (we did not use the ADM model from compute_dire.py because the pre-trained model for ADM was not open-source at the time we started our research). In calculating DIRE, we normalize both the original and reconstructed images to the (-1, 1) range and then compute the absolute error to obtain features for training. We only reported our actual reproduction results (acknowledging that there might be shortcomings in our reproduction). If you can reproduce higher scores, you are welcome to report your results as well. Lastly, we greatly appreciate you pointing out the shortcomings in our work.

**Comment by Ck2125** [comment]
Thank you very much for your prompt reply. I have been working on reproducing the DIRE paper on the Genimage dataset, and I encountered an issue that I would like to discuss. When testing the checkpoint trained on sdv4 (with settings matching those in Table 2 of your paper), I observed that the model achieved an accuracy close to 1 across the eight subclasses of Genimage.

Upon further investigation, I came across a discussion on https://github.com/ZhendongWang6/DIRE/issues/11 suggesting that, in the DIRE dataset, the DIRE images corresponding to real images are in .JPEG format, while those for fake images are in .PNG format. It seems that the checkpoint provided by DIRE might primarily recognize JPEG artifacts, rather than detecting fake images. Since the image format of the Genimage dataset follows the same convention, this could explain why the model trained using compute_dire.py and the ADM checkpoint from DIRE yields an unusually high accuracy on the eight subclasses of Genimage—if the model is only identifying JPEG artifacts.

However, I noticed that in your paper, Table 2 shows that DIRE achieved high accuracy on the sdv4, sdv5, and wukong subclasses, but not on the other five subclasses. I am curious whether your experiments at that time solved the problems with the open source model and checkpoints in the original DIRE paper that I mentioned above.

Additionally, I would be grateful if you could share the complete code for reconstructing images (and the corresponding checkpoints for the reconstruction model), as well as the process for calculating DIRE graphs. If it is not convenient to make the code public, would it be possible for you to send it to my email at 2480178180@qq.com? I would sincerely appreciate your help!

**Comment by Ck2125** [comment]
Thank you very much for your prompt reply. I have been working on reproducing the DIRE paper on the Genimage dataset, and I encountered an issue that I would like to discuss. When testing the checkpoint trained on sdv4 (with settings matching those in Table 2 of your paper), I observed that the model achieved an accuracy close to 1 across the eight subclasses of Genimage.

Upon further investigation, I came across a discussion on https://github.com/ZhendongWang6/DIRE/issues/11 suggesting that, in the DIRE dataset, the DIRE images corresponding to real images are in .JPEG format, while those for fake images are in .PNG format. It seems that the checkpoint provided by DIRE might primarily recognize JPEG artifacts, rather than detecting fake images. Since the image format of the Genimage dataset follows the same convention, this could explain why the model trained using compute_dire.py and the ADM checkpoint from DIRE yields an unusually high accuracy on the eight subclasses of Genimage—if the model is only identifying JPEG artifacts.

However, I noticed that in your paper, Table 2 shows that DIRE achieved high accuracy on the sdv4, sdv5, and wukong subclasses, but not on the other five subclasses. I am curious whether your experiments at that time solved the problems with the open source model and checkpoints in the original DIRE paper that I mentioned above.

Additionally, I would be grateful if you could share the complete code for reconstructing images (and the corresponding checkpoints for the reconstruction model), as well as the process for calculating DIRE graphs. If it is not convenient to make the code public, would it be possible for you to send it to my email at [2480178180@qq.com](mailto:2480178180@qq.com)? I would sincerely appreciate your help!

**Comment by Ck2125** [comment]
> Thank you very much for your prompt reply. I have been working on reproducing the DIRE paper on the Genimage dataset, and I encountered an issue that I would like to discuss. When testing the checkpoint trained on sdv4 (with settings matching those in Table 2 of your paper), I observed that the model achieved an accuracy close to 1 across the eight subclasses of Genimage.
> 
> Upon further investigation, I came across a discussion on [ZhendongWang6/DIRE#11](https://github.com/ZhendongWang6/DIRE/issues/11) suggesting that, in the DIRE dataset, the DIRE images corresponding to real images are in .JPEG format, while those for fake images are in .PNG format. It seems that the checkpoint provided by DIRE might primarily recognize JPEG artifacts, rather than detecting fake images. Since the image format of the Genimage dataset follows the same convention, this could explain why the model trained using compute_dire.py and the ADM checkpoint from DIRE yields an unusually high accuracy on the eight subclasses of Genimage—if the model is only identifying JPEG artifacts.
> 
> However, I noticed that in your paper, Table 2 shows that DIRE achieved high accuracy on the sdv4, sdv5, and wukong subclasses, but not on the other five subclasses. I am curious whether your experiments at that time solved the problems with the open source model and checkpoints in the original DIRE paper that I mentioned above.
> 
> Additionally, I would be grateful if you could share the complete code for reconstructing images (and the corresponding checkpoints for the reconstruction model), as well as the process for calculating DIRE graphs. If it is not convenient to make the code public, would it be possible for you to send it to my email at [2480178180@qq.com](mailto:2480178180@qq.com)? I would sincerely appreciate your help!

Sorry, I just saw diffusion_rec.py in your open source code. Is this the code you use to reconstruct the image?

**Comment by beibuwandeluori** [comment]
Yes, my code for reconstructing images is in "data/diffusion_rec.py", and the code for calculating DIRE is in the "calculate_sdie" function within "data/dataset.py".

**Comment by beibuwandeluori** [comment]
However, I noticed that in your paper, Table 2 shows that DIRE achieved high accuracy on the sdv4, sdv5, and wukong subclasses, but not on the other five subclasses. I am curious whether your experiments at that time solved the problems with the open source model and checkpoints in the original DIRE paper that I mentioned above.
关于这个问题回复（看你使用qq邮箱，直接中文回复了）：因为DIRE论文没有在GenImage上的实验，所以我们当时只是按照DIRE的论文公式复现出GenImage上的结果并如实报道到我们论文中，并没特别关注为何在除了sdv4, sdv5, 和wukong 外其他几个子类性能下降，我们主要是关注如何去优化我们的DRCT算法。

**Comment by Ck2125** [comment]
好的，十分感谢你非常详细和快速的回复！！！很少见到作者在github上这么快速且耐心地回复问题。

**Comment by Ck2125** [comment]
你好，想再问一下：我用你们开源的重建与计算dire图的代码，在sdv4的训练集的部分图片上进行了测试，结果发现得到的dire图近似于全黑（有部分得到的dire图是全黑，大部分接近全黑），和DIRE原论文中展现的dire图（[https://github.com/ZhendongWang6/DIRE/blob/main/figs/dire.png）差距很大。
虽然我知道两篇论文的reconstruction model并不相同，还是想问一下你当时进行这部分实验时，是出现这样的情况吗？

**Comment by Ck2125** [comment]
你好，想再问一下：我用你们开源的重建与计算dire图的代码，在sdv4的训练集的部分图片上进行了测试，结果发现得到的dire图近似于全黑（有部分得到的dire图是全黑，大部分接近全黑），和DIRE原论文中展现的dire图（[https://github.com/ZhendongWang6/DIRE/blob/main/figs/dire.png]）差距很大。
虽然我知道两篇论文的reconstruction model并不相同，还是想问一下你当时进行这部分实验时，是出现这样的情况吗？

**Comment by Ck2125** [comment]
明白，感谢你的回复。另外还想问一下：因为重建确实如你在https://github.com/beibuwandeluori/DRCT/issues/12提到的十分耗时，想问一下你愿不愿意公开一下在你论文中表2在genimage数据集上训练并测试时，DIRE的八个子类验证集在测试时的average precision？如果你愿意公开这些数据，我们将不胜感激。

**Comment by Ck2125** [comment]
因为重建确实如你在(https://github.com/beibuwandeluori/DRCT/issues/12) 提到的十分耗时，想问一下你愿不愿意公开一下在你论文中表2在genimage数据集上训练并测试时，DIRE的八个子类验证集在测试时的average precision？如果您愿意公开这些数据，我们将不胜感激。

**Comment by beibuwandeluori** [comment]
from sklearn.metrics import average_precision_score
# 假设这是你的真实标签（1表示AI生成的，0表示真实的）
y_true = [0, 1, 1, 0, 1, 0, 1]
# 假设这是你的模型预测的概率或得分（越接近1表示越可能是AI生成的）
y_scores = [0.1, 0.4, 0.35, 0.8, 0.7, 0.2, 0.9]
ap_score = average_precision_score(y_true, y_scores)
我们之前并没有计算过average precision，如果average precision是按照上面方式计算，我大概在一周后左右给你结果（最近比较忙），如果上面的计算例子不对，请你提供代码样例出来

**Comment by Ck2125** [comment]
> ```
> from sklearn.metrics import average_precision_score
> # 假设这是你的真实标签（1表示AI生成的，0表示真实的）
> y_true = [0, 1, 1, 0, 1, 0, 1]
> # 假设这是你的模型预测的概率或得分（越接近1表示越可能是AI生成的）
> y_scores = [0.1, 0.4, 0.35, 0.8, 0.7, 0.2, 0.9]
> ap_score = average_precision_score(y_true, y_scores)
> ```
> 
> 我们之前并没有计算过average precision，如果average precision是按照上面方式计算，我大概在一周后左右给你结果（最近比较忙），如果上面的计算例子不对，请你提供代码样例出来

就是这么计算，十分感谢！

**Comment by Ck2125** [comment]
> ![Image](https://github.com/user-attachments/assets/e1779b0c-b679-4a15-89d2-9d549b94b468) 这个是我重跑计算的average precision结果，希望对你有帮助！

非常感谢你的帮助！

**Comment by hjs-2002** [comment]
> 你好，想再问一下：我用你们开源的重建与计算dire图的代码，在sdv4的训练集的部分图片上进行了测试，结果发现得到的dire图近似于全黑（有部分得到的dire图是全黑，大部分接近全黑），和DIRE原论文中展现的dire图（[https://github.com/ZhendongWang6/DIRE/blob/main/figs/dire.png]）差距很大。 虽然我知道两篇论文的reconstruction model并不相同，还是想问一下你当时进行这部分实验时，是出现这样的情况吗？

您好，能不能提供相关重建图片dire的代码啊，我参考学习一下，感觉我的好像有些问题
